src/bootstrap/step.rs

```diff
@@ -838,14 +838,22 @@ invalid rule dependency graph detected, was a rule added and maybe typo'd?
             } else {
                 &self.build.config.target
             };
-            // If --target was specified but --host wasn't specified, don't run
-            // any host-only tests
+            // Determine the actual targets participating in this rule.
             let arr = if rule.host {
-                if self.build.flags.target.len() > 0 &&
-                   self.build.flags.host.len() == 0 {
-                    &hosts[..0]
+                if kind == Kind::Dist {
+                    // Keep the full projection from build triple to the hosts
+                    // for the dist steps, now that the hosts array is truncated
+                    // to avoid duplication of work.
+                    &self.build.config.host
```

Comment:
Ah I think we can avoid the check for `Dist` entirely during the clause for `--target`. We basically just need to filter down an array of `hosts` without the `Dist` clause